src/relax/transform/lambda_lift.cc

```diff
@@ -336,176 +266,235 @@ class LambdaLifter : public ExprMutator {
-          if (const auto* var = seq_expr_node->body.as<VarNode>()) {
-            return HasClosure(GetRef<Var>(var));
+  Expr VisitExpr_(const CallNode* call_node) final {
+    auto call = GetRef<Call>(call_node);
+
+    auto orig_sinfo = Downcast<StructInfo>(call->struct_info_);
+
+    if (auto opt_var = call->op.as<Var>()) {
+      auto var = opt_var.value();
+
+      // Call "relax.invoke_closure" to invoke closure
+
+      if (bool is_closure = IsClosure(var);
+          is_closure && builder_->LookupBinding(var).as<CallNode>()) {
+        // if the original op was pure, we should use invoke_pure_closure
+        Call orig_call = Downcast<Call>(builder_->LookupBinding(var));
+        bool is_pure = [&]() -> bool {
+          if (auto op = orig_call->op.as<Op>()) {
+            static const auto& purity_map = Op::GetAttrMap<Bool>("FPurity");
+            return purity_map.get(op.value(), Bool(false))->value;
+          } else if (const auto* func_sinfo =
+                         orig_call->op->struct_info_.as<FuncStructInfoNode>()) {
+            return func_sinfo->purity;
+          } else {
+            LOG(FATAL) << "Could not determine purity of call to " << orig_call->op
+                       << ", as it is neither a tvm::Op (type = \"" << orig_call->op->GetTypeKey()
+                       << "\"), "
+                       << "nor is is annotated with FuncStructInfo (sinfo = "
+                       << orig_call->op->struct_info_ << ")";
           }
-        }
+        }();
+
+        auto prev = call;
+        call = Call(is_pure ? invoke_pure_closure_op_ : invoke_closure_op_,
+                    {var, Tuple(call->args)}, {}, {orig_sinfo});
       }
-    } else if (const auto* func_node = val.as<FunctionNode>()) {
-      if (const auto* call_node = func_node->body.as<CallNode>()) {
-        if (call_node->op == make_closure_op_) {
-          return true;
+    }
+
+    if (auto opt_var = call->op.as<Var>()) {
+      auto var = opt_var.value();
+      if (auto it = nested_closure_map_.find(var); it != nested_closure_map_.end()) {
+        Call nested_call = it->second;
+
+        Array<relay::Expr> new_args = call->args;
+        for (const auto arg : nested_call->args) {
+          new_args.push_back(arg);
         }
+
+        auto prev = call;
+        call = Call(nested_call->op, new_args, call->attrs, call->sinfo_args);
       }
-    } else if (const auto* call_node = val.as<relax::CallNode>()) {
+    }
+
+    return ExprMutator::VisitExpr_(call.get());
+  }
+
+  Expr VisitExpr_(const VarNode* op) override {
+    auto var = GetRef<Var>(op);
+    if (auto it = rebind_map_.find(var); it != rebind_map_.end()) {
+      return it->second;
+    }
+    return ExprMutator::VisitExpr_(op);
+  }
+
+  bool IsClosure(Expr val) {
+    static int depth = -1;
+    struct Context {
+      explicit Context(int* ptr) : ptr(ptr) { (*ptr)++; }
+      ~Context() { (*ptr)--; }
+      int* ptr;
+    } context(&depth);
```

Comment:
This doesn't appear to be used anywhere.

Reply:
Whoops, debug code for indenting prints during nested calls.  Removed.